### PR TITLE
Support binary operation with Timedelta in Pyspark backend

### PR DESCRIPTION
**Pull request by emilyreff7**
This PR adds support in the Pyspark backend for doing binary operations (e.g. add/subtract) on timestamp columns using Pandas Timedeltas, for example:

```
dt['timestamp_col'] + pd.Timedelta(days=5)
```

### CODE REVIEW

File: ibis/backends/pyspark/compiler.py [hunk 1]
```diff
@@ -1532,8 +1543,15 @@ def _get_interval_col(
             'Interval unit "{}" is not allowed. Allowed units are: '
             '{}'.format(dtype.unit, allowed_units)
         )
+
+    if isinstance(op.value, pd.Timedelta):
+        pandas_unit = _time_unit_mapping_pd_timedelta[dtype.unit]
```
Comment (icexelloss) [inline ibis/backends/pyspark/compiler.py:1548]: I am not sure we need this. Does pyspark not work with pd.Timedelta directly?

Comment (emilyreff7) [reply ibis/backends/pyspark/compiler.py:1548]: If we just pass in op.value as the existing code does, then Pyspark blows up with a ParseException, e.g. if we run 

```
t.timestamp_col + pd.Timedelta('10s')
```
the error is

```
E   pyspark.sql.utils.ParseException: 
E   mismatched input ':' expecting {'ADD', 'AFTER', 'ALL', 'ALTER'...
<truncated>
E   
E   == SQL ==
E   INTERVAL 0 days 00:00:10 second
E   ------------------^^^

```
so at least if we want to use this `INTERVAL` syntax, Pyspark doesn't do well with the colons in the Pandas Timedelta representation.

File: ibis/backends/pyspark/compiler.py [hunk 2]
```diff
@@ -1532,8 +1543,15 @@ def _get_interval_col(
...
+        value = getattr(op.value.components, pandas_unit)
+    else:
+        value = op.value
```
Comment (icexelloss) [inline ibis/backends/pyspark/compiler.py:1551]: What hits the else here? (What other than pd.Timedelta supported)?

File: ibis/backends/pyspark/compiler.py [hunk 3]
```diff
@@ -1532,8 +1543,15 @@ def _get_interval_col(
             'Interval unit "{}" is not allowed. Allowed units are: '
             '{}'.format(dtype.unit, allowed_units)
         )
+
+    if isinstance(op.value, pd.Timedelta):
+        pandas_unit = _time_unit_mapping_pd_timedelta[dtype.unit]
+        value = getattr(op.value.components, pandas_unit)
```
Comment (icexelloss) [inline ibis/backends/pyspark/compiler.py:1549]: Is this safe? Can there be more than one components? e.g., hour and minute?

Comment (icexelloss) [reply ibis/backends/pyspark/compiler.py:1549]: I think it is probably better to do sth like this

```
td = op.value # pd.Timedelta
td_nanos = td.value
# Spark supports micro as the smallest unit, so if we have any nanos component, raise Error

td_micro = td_nanos / 1000

spark_inteval = F.expr(f'INTERVAL {td_mirco} MICROSECOND')
```

Comment (icexelloss) [reply ibis/backends/pyspark/compiler.py:1549]: This way you don't even need to keep the mapping

File: ibis/backends/pyspark/compiler.py (hunk 1 above)
Comment (icexelloss) [reply ibis/backends/pyspark/compiler.py:1548]: kk

**Review by icexelloss — CHANGES_REQUESTED** [review]
Request for small change to simply the logic

File: ibis/backends/tests/test_temporal.py [hunk 4]
```diff
@@ -306,6 +307,29 @@ def test_temporal_binop(backend, con, alltypes, df, expr_fn, expected_fn):
     backend.assert_series_equal(result, expected)
 
 
+@pytest.mark.parametrize(
+    ('timedelta', 'temporal_fn'),
+    itertools.product(
+        ['5W', '3d', '2h', '3m', '10s'],
```
Comment (icexelloss) [inline ibis/backends/tests/test_temporal.py:313]: Please also test really large timedelta like 100 years to make sure we don't overflow

Comment (icexelloss) [reply ibis/backends/tests/test_temporal.py:313]: Please add also test for unsupported shifts (e.g. 1 nanos)

File: ibis/backends/pyspark/compiler.py (hunk 2 above)
Comment (emilyreff7) [reply ibis/backends/pyspark/compiler.py:1551]: All of these existing tests for example hit the else: https://github.com/ibis-project/ibis/blob/master/ibis/backends/tests/test_temporal.py#L299
`ibis.interval`, `ibis.date`, etc.

File: ibis/backends/pyspark/compiler.py (hunk 3 above)
Comment (emilyreff7) [reply ibis/backends/pyspark/compiler.py]: Makes sense, definitely simpler than maintaining a mapping - and Pandas errors out immediately anyway if you pass a value to Timedelta that would cause an overflow, e.g. 

```
>>> pd.Timedelta('1000Y')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/_libs/tslibs/timedeltas.pyx", line 1217, in pandas._libs.tslibs.timedeltas.Timedelta.__new__
  File "pandas/_libs/tslibs/timedeltas.pyx", line 462, in pandas._libs.tslibs.timedeltas.parse_timedelta_string
  File "pandas/_libs/tslibs/timedeltas.pyx", line 514, in pandas._libs.tslibs.timedeltas.timedelta_from_spec
  File "pandas/_libs/tslibs/timedeltas.pyx", line 312, in pandas._libs.tslibs.timedeltas.cast_from_unit
OverflowError: Python int too large to convert to C long
```

File: ibis/backends/tests/test_temporal.py (hunk 4 above)
Comment (emilyreff7) [reply ibis/backends/tests/test_temporal.py]: Added test for both

**Review by icexelloss — COMMENTED** [review]
LGTM. +1

**Review by icexelloss — APPROVED** [review]

File: ibis/backends/pyspark/compiler.py (hunk 3 above)
Comment (icexelloss) [reply ibis/backends/pyspark/compiler.py]: Great!

**Comment by jreback** [comment]
thanks @emilyreff7
